File: backend/app/routers/domains.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from app.database import get_db
from app.models.domain import Domain
from app.models.client import Client
from app.schemas import DomainCreate, DomainUpdate, DomainRead

router = APIRouter(prefix="/domains", tags=["domains"])


@router.get("", response_model=list[DomainRead])
def get_domains(db: Session = Depends(get_db)):
    return db.query(Domain).order_by(Domain.id).all()


@router.get("/{domain_id}", response_model=DomainRead)
def get_domain(domain_id: int, db: Session = Depends(get_db)):
    domain = db.query(Domain).filter(Domain.id == domain_id).first()
    if not domain:
        raise HTTPException(status_code=404, detail="Domain not found")
    return domain


@router.post("", response_model=DomainRead, status_code=status.HTTP_201_CREATED)
def create_domain(payload: DomainCreate, db: Session = Depends(get_db)):
    client = db.query(Client).filter(Client.id == payload.client_id).first()
    if not client:
        raise HTTPException(status_code=404, detail="Client not found")

    domain = Domain(**payload.model_dump())
    db.add(domain)
    db.commit()
    db.refresh(domain)
    return domain


@router.put("/{domain_id}", response_model=DomainRead)
def update_domain(domain_id: int, payload: DomainUpdate, db: Session = Depends(get_db)):
    domain = db.query(Domain).filter(Domain.id == domain_id).first()
    if not domain:
        raise HTTPException(status_code=404, detail="Domain not found")

    client = db.query(Client).filter(Client.id == payload.client_id).first()
    if not client:
        raise HTTPException(status_code=404, detail="Client not found")

    for key, value in payload.model_dump().items():
        setattr(domain, key, value)

    db.commit()
    db.refresh(domain)
    return domain


@router.delete("/{domain_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_domain(domain_id: int, db: Session = Depends(get_db)):
    domain = db.query(Domain).filter(Domain.id == domain_id).first()
    if not domain:
        raise HTTPException(status_code=404, detail="Domain not found")

    db.delete(domain)
    db.commit()

File: frontend/src/api/domains.js
import apiClient from "./client";

export async function getDomains(contractId) {
  const params = {};

  if (contractId !== undefined && contractId !== null && contractId !== "") {
    params.contract_id = contractId;
  }

  const response = await apiClient.get("/domains/", { params });
  return response.data;
}

export async function getMyDomains() {
  const response = await apiClient.get("/domains/my");
  return response.data;
}

export async function createDomain(domainData) {
  const response = await apiClient.post("/domains/", domainData);
  return response.data;
}

export async function updateDomain(domainId, domainData) {
  const response = await apiClient.put(`/domains/${domainId}`, domainData);
  return response.data;
}

export async function deleteDomain(domainId) {
  const response = await apiClient.delete(`/domains/${domainId}`);
  return response.data;
}
